src/Microsoft.AspNetCore.Routing/Tree/TreeRouteBuilder.cs

```diff
@@ -221,6 +272,36 @@ private void AddEntryToTree(UrlMatchingTree tree, InboundRouteEntry entry)
         {
             var current = tree.Root;
 
+            /* The url matching tree represents all the routes asociated with a given
+             * order. Each node in the tree represents all the different categories
+             * a segment can have for which there is a defined inbound route entry.
+             * Each node contains a set of Matches that indicate all the routes for which
+             * a URL is a potential match. This list contains the routes with the same
+             * number of segments and the routes with the same number of segments plus an
+             * additional catch all parameter (as it can be empty).
+             * For example, for a set of routes like:
+             * 'Customer/Index/{id}'
+             * '{Controller}/{Action}/{*parameters}'
+             * 
+             * The route tree will look like:
+             * Root ->
+             *     Literals: Customer ->
+             *                   Literals: Index ->
+             *                                Parameters: {id}
+             *                                                Matches: 'Customer/Index/{id}'
+             *     Parameters: {Controller} ->
+             *                     Parameters: {Action} ->
+             *                                     Matches: '{Controller}/{Action}/{*parameters}'
+             *                                     CatchAlls: {*parameters}
+             *                                                    Matches: '{Controller}/{Action}/{*parameters}'
+             * 
+             * When the tree router tries to match a route, it iterates the list of url matching trees
+             * in ascending order. For each tree it traverses each node starting from the root in the
+             * following order: Literals, constrained parameters, parameters, constrained catch all routes, catch alls.
+             * When it gets to a node of the same length as the route its trying to match, it simply looks at the list of
+             * candidates (which is in precence order) and tries to match the url against it.
+             */
+
```

Comment:
I've tried to explain the tree construction to the best of my knowledge. @rynowak if you can express it better or in a more accurate way, just use this comment as a base and tell me what to replace it with.